internal/btf/btf.go

```diff
@@ -147,14 +147,134 @@ func loadSpecFromELF(file *internal.SafeELFFile, variableOffsets map[variable]ui
 		return spec, nil
 	}
 
-	spec.funcInfos, spec.lineInfos, spec.coreRelos, err = parseExtInfos(btfExtSection, file.ByteOrder, spec.strings)
-	if err != nil {
-		return nil, fmt.Errorf("can't read ext info: %w", err)
+	if err := spec.loadExtInfos(btfExtSection); err != nil {
+		return nil, fmt.Errorf("can't parse ext info: %w", err)
 	}
 
 	return spec, nil
 }
 
+// loadExtInfos parses the .BTF.ext section and populates spec's funcInfo,
+// lineInfo and coreRelos fields using the contents.
+func (spec *Spec) loadExtInfos(sec *elf.Section) error {
+	// Open unbuffered section reader. binary.Read() calls io.ReadFull on
+	// the header structs, resulting in one syscall per header.
+	r := sec.Open()
+
+	extHeader, err := parseBTFExtHeader(r, spec.byteOrder)
+	if err != nil {
+		return fmt.Errorf("parsing BTF extension header: %w", err)
+	}
+
+	coreHeader, err := parseBTFExtCoreHeader(r, spec.byteOrder, extHeader)
+	if err != nil {
+		return fmt.Errorf("parsing BTF CO-RE header: %w", err)
+	}
+
+	buf := internal.NewBufferedSectionReader(sec, extHeader.funcInfoStart(), int64(extHeader.FuncInfoLen))
+	funcInfos, err := parseFuncInfos(buf, spec.byteOrder, spec.strings)
+	if err != nil {
+		return fmt.Errorf("parsing BTF function info: %w", err)
+	}
+
+	buf = internal.NewBufferedSectionReader(sec, extHeader.lineInfoStart(), int64(extHeader.LineInfoLen))
+	lineInfos, err := parseLineInfos(buf, spec.byteOrder, spec.strings)
+	if err != nil {
+		return fmt.Errorf("parsing BTF line info: %w", err)
+	}
+
+	relos := make(map[string]CoreRelos)
+	if coreHeader != nil && coreHeader.CoreReloOff > 0 && coreHeader.CoreReloLen > 0 {
+		buf = internal.NewBufferedSectionReader(sec, extHeader.coreReloStart(coreHeader), int64(coreHeader.CoreReloLen))
+		relos, err = parseCoreRelos(buf, spec.byteOrder, spec.strings)
+		if err != nil {
+			return fmt.Errorf("parsing CO-RE relocation info: %w", err)
+		}
+	}
+
+	if err := spec.linkExtInfos(funcInfos, lineInfos, relos); err != nil {
+		return fmt.Errorf("linking funcInfos and lineInfos: %w", err)
+	}
+
+	return nil
+}
+
+// linkExtInfos takes FuncInfos, LineInfos and CoreRelos indexed by section and
+// transforms them to be indexed by function. Retrieves function names from
+// the BTF spec.
+func (spec *Spec) linkExtInfos(sfi map[string]FuncInfos, sli map[string]LineInfos, scr map[string]CoreRelos) error {
```

Comment:
> The only reason to be on spec is to set fields in spec, which you can do in the caller.

But why? It's neater having the fields set as part of the function call, it's all wrapped up nicely. Setting fields isn't the only thing `spec` is used for. If we need to keep `spec` around as an argument, it might as well be a receiver..

> Rationale: I look at this signature and can't get an idea for what this does.

I've read something a while back that if you use the word 'and' in a function docstring, the function already does too much and should probably be split up. :) I think your statement applies to many functions around the codebase, doubly so if they don't have docstrings (which many of them don't).

I think the docstring is abundantly clear, but calling it `splitExtInfos` would make more sense, agreed.